### PR TITLE
i/apparmor,s/apparmor: move snap-confine snippet generation to sandbox

### PR DESCRIPTION
**Pull request by Meulengracht**
Refactoring part, moving snippet generation from interface/apparmor to sandbox/apparmor. Extracted from PR https://github.com/snapcore/snapd/pull/12288. This will be extended to handle homedirs in a seperate PR.

### CODE REVIEW

**Review by mvo5 — APPROVED** [review]
Thank for this refactor

File: sandbox/apparmor/profile.go [hunk 1]
```diff
@@ -251,3 +286,68 @@ var SnapConfineDistroProfilePath = func() string {
...
+
+	// Ensure that generated policy is what we computed above.
+	created, removed, err := osutil.EnsureDirState(dirs.SnapConfineAppArmorDir, glob, policy)
+	if err != nil {
+		return false, fmt.Errorf("cannot synchronize snap-confine policy: %s", err)
+	}
+	return len(created) > 0 || len(removed) > 0, nil
+}
+
+func RemoveSnapConfineSnippets() error {
```
Comment (pedronis) [inline sandbox/apparmor/profile.go:349]: this should have a doc comment and could use a unit test in the package

File: sandbox/apparmor/profile.go [hunk 2]
```diff
@@ -251,3 +286,68 @@ var SnapConfineDistroProfilePath = func() string {
...
+// SetupSnapConfineSnippets inspects the system and sets up local apparmor
+// policy for snap-confine. Local policy is included by the system-wide policy.
+// Returns whether any modifications was made to the snap-confine snippets.
+func SetupSnapConfineSnippets() (wasChanged bool, err error) {
+	// Create the local policy directory if it is not there.
+	if err := os.MkdirAll(dirs.SnapConfineAppArmorDir, 0755); err != nil {
+		return false, fmt.Errorf("cannot create snap-confine policy directory: %s", err)
+	}
+
+	// Location of the generated policy.
```
Comment (pedronis) [inline sandbox/apparmor/profile.go:299]: this comment wasn't relevant anymore, the glob is not a location??

File: sandbox/apparmor/profile.go [hunk 3]
```diff
@@ -251,3 +286,68 @@ var SnapConfineDistroProfilePath = func() string {
...
+	// Ensure that generated policy is what we computed above.
+	created, removed, err := osutil.EnsureDirState(dirs.SnapConfineAppArmorDir, glob, policy)
+	if err != nil {
+		return false, fmt.Errorf("cannot synchronize snap-confine policy: %s", err)
+	}
+	return len(created) > 0 || len(removed) > 0, nil
+}
+
+func RemoveSnapConfineSnippets() error {
+	glob := "*"
```
Comment (pedronis) [inline sandbox/apparmor/profile.go:350]: this could be a const or just hard-coded in the call

File: sandbox/apparmor/profile.go [hunk 4]
```diff
@@ -251,3 +286,68 @@ var SnapConfineDistroProfilePath = func() string {
...
+// policy for snap-confine. Local policy is included by the system-wide policy.
+// Returns whether any modifications was made to the snap-confine snippets.
+func SetupSnapConfineSnippets() (wasChanged bool, err error) {
+	// Create the local policy directory if it is not there.
+	if err := os.MkdirAll(dirs.SnapConfineAppArmorDir, 0755); err != nil {
+		return false, fmt.Errorf("cannot create snap-confine policy directory: %s", err)
+	}
+
+	// Location of the generated policy.
+	glob := "*"
```
Comment (pedronis) [inline sandbox/apparmor/profile.go:300]: this can be a const now, otoh maybe it can just be hard-coded below?

**Review by pedronis — COMMENTED** [review]
some small comments

File: sandbox/apparmor/profile.go (hunk 2 above)
Comment (Meulengracht) [reply sandbox/apparmor/profile.go:299]: Good point, removed the comment as it had no place there, and wasn't needed after removing the `glob` variable

File: sandbox/apparmor/profile.go (hunk 3 above)
Comment (Meulengracht) [reply sandbox/apparmor/profile.go:350]: Good point!

File: sandbox/apparmor/profile.go (hunk 1 above)
Comment (Meulengracht) [reply sandbox/apparmor/profile.go:349]: Fixed!

File: sandbox/apparmor/profile.go [hunk 5]
```diff
@@ -339,15 +337,16 @@ func SetupSnapConfineSnippets() (wasChanged bool, err error) {
...
-	created, removed, err := osutil.EnsureDirState(dirs.SnapConfineAppArmorDir, glob, policy)
+	created, removed, err := osutil.EnsureDirState(dirs.SnapConfineAppArmorDir, "*", policy)
 	if err != nil {
 		return false, fmt.Errorf("cannot synchronize snap-confine policy: %s", err)
 	}
 	return len(created) > 0 || len(removed) > 0, nil
 }
 
+// RemoveSnapConfineSnippets clears out any previously written apparmor snippets
+// for snap-confine
```
Comment (pedronis) [inline sandbox/apparmor/profile.go:348]: missing . at the end

**Review by pedronis — APPROVED** [review]
thank you, small detail

File: sandbox/apparmor/profile.go (hunk 5 above)
Comment (Meulengracht) [reply sandbox/apparmor/profile.go]: thank you for catching this!

**Pull request merged**